### DISCUSSION TITLE
Using PROXY Protocol on Technitium fronted with a layer-4 router

### DISCUSSION THREAD

**Discussion by stratself** [TechnitiumSoftware/DnsServer#1099] [Show and tell]
Just wanna share my configs for routing packets with PROXY Protocol enabled to Technitium. This helps with retrieving source IPs of DNS queries when Technitium can't do so natively (e.g. behind NAT). Also available as [blog](https://muoi.me/~stratself/articles/nginx-dns-proxy-protocol/).

```mermaid
graph LR
    user -->|"53/udp"| nginx -->|"538/udp"|technitium
    user -->|"53/tcp"| nginx -->|"538/tcp"| technitium
```

**1. Enable the protocols**

Technitium Homepage -> Settings -> Optional Protocols -> check on **Enable DNS-over-TCP-PROXY** and **Enable DNS-over-UDP-PROXY**. By default these ports are `538/tcp` and `538/udp`, but you can change it if you want.

If you're using Technitium in a container, make sure to expose these ports as well.

**2. Configure a nginx service**

Install nginx via container or on host. Add this/mount this to your `/etc/nginx/nginx.conf`.

```
worker_processes  auto;

error_log  /var/log/nginx/error.log notice;
pid        /tmp/nginx.pid;


events {
    # increased from 1024 to avoid congestions
    worker_connections  4096;
}

# use the stream directive
stream {
    # for DNS-over-UDP-PROXY
    server {
        listen              53 udp;
        proxy_pass          127.0.0.1:538;
        proxy_protocol      on;
    }
    server {
    # for DNS-over-TCP-PROXY
        listen              53;
        proxy_pass          127.0.0.1:538;
        proxy_protocol      on;
    }
}
```

If nginx is on a different machine from Technitium, replace `127.0.0.1` with appropriate Technitium's IP address.

Enable/Restart nginx.

**3. Test the services**

Using [dog](https://dns.lookup.dog/) these should work:

```sh
dog -U example.com @<your-nginx-machine-ip>:53 # UDP
dog -T example.com @<your-nginx-machine-ip>:53 # TCP
```

And the correct client IP should be logged in Technitium.

---

PROXY Protocol helps with workarounds to some site-to-site DNS scenarios, which would make Split Horizon feasible. As per the docs, the same method can also be used for DoT as well.

```mermaid
graph LR
    user -->|"853/tcp"| nginx["TCP router with TLS termination"] -->|"538/tcp"|technitium
```

**4. Extra tuning**

As mentioned in [a comment below](https://github.com/TechnitiumSoftware/DnsServer/discussions/1099#discussioncomment-14840548), you can append `reuseport` to the listen line. This should allow nginx's workers to share the same port 53 and increase throughput and performance of your DNS proxy.

---

Please correct and suggest improvements. Also please share configs for other routers/reverse proxies e.g. **Caddy-L4** or **HAProxy** (Nginx is the only one I found working so far).

**Comment by ShreyasZare** [comment]
Thanks for posting this here.

**Comment by stratself** [comment]
Managed to spin up an equivalent [Caddy-L4](https://github.com/mholt/caddy-l4/) config, for anybody interested:

```Caddyfile
{
    debug
    layer4 {
        udp/:53 {
            @udns dns
			route @udns {
				proxy {
                    proxy_protocol v2
                    upstream udp/127.0.0.1:538
                }
            }
        }
        tcp/:53 {
			@tdns dns
			route @tdns {
				proxy {
                    proxy_protocol v2
                    upstream tcp/127.0.0.1:538
                }
            }
        }
    }
}
```

Huge thanks to people who resolved https://github.com/mholt/caddy-l4/issues/269

**Reply by stratself** [reply]
Note: the `dns` matcher in Caddy doesn't seem to support Zone Transfer (XFR) messages. When using XFR-over-TCP, an `EndOfStreamException` error is generated (also see #1189).

Disabling the matcher rules for :53/tcp (and therefore route every TCP packet to Technitium) should make it work again:


```
tcp/:53 {
    route {
        proxy {
            proxy_protocol v2
            upstream tcp/127.0.0.1:538
            }
        }
    }
```

**Comment by bcookatpcsd** [comment]
dnsdist.. 

```
newServer({ address="[::1]:538", useProxyProtocol=true, tcpFastOpen=true, healthCheckMode='lazy', name="pxy-T
echnitium", checkName="accounts.google.com", sockets="4", pool="pc-proxyv2" })
```

YMMV 

<img width="1845" height="424" alt="image" src="https://github.com/user-attachments/assets/d8c6be72-aefa-49b3-ac3d-0a86829751a5" />

**Reply by aberthil** [reply]
Just to double check you really just have to point the port 538 for the backend server with proxy server enabled and you can see all client ip on the dns, correct?

**Reply by aberthil** [reply]
No need to even have the ngix server here right?

**Comment by kinslayer1982** [comment]
The first diagram is exactly what I came up with. I ditched my Pihole because it does not support the proxy protocol and so it only sees the nginx ip, not the clients. But after setting it up like this my logs get flooded with
```
[UDPPROXY] System.IO.InvalidDataException: The stream does not contain PROXY protocol header.
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.CreateAsServerAsync(Stream baseStream, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\ProxyProtocol\ProxyProtocolStream.cs:line 138
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 414
```
messages.
Is this normal behavior? DNS resolution seems to be working.

**Reply by ShreyasZare** [reply]
Thanks for the details. It is not normal and error indicates that UDP requests are being received without the PROXY protocol enabled. Just make sure that the PROXY protocol is enabled in your nginx config and also that no other DNS requests are being forwarded to the 538 port on DNS server.

**Reply by kinslayer1982** [reply]
All these messages contain the ip of nginx reverse proxy. Except for nginx nothing is sending anything to this port and `proxy_protocol on;` is set in the stream server block.

I have tried with different nginx versions with same results. I thought it might be expected behavior because I've read about problems with upd and proxy protocol headers with other load balancers/reverse proxies.

**Reply by ShreyasZare** [reply]
Please share your nginx config here so that I can understand it. When this feature was implemented, it was tested with nginx and it worked well without any of these error messages.

**Comment by kinslayer1982** [comment]
This is the nginx conf:
```
user  nginx;
worker_processes  auto;

error_log  /var/log/nginx/error.log warn;
pid        /var/run/nginx.pid;

events {
    worker_connections  4096;
}


stream {
    include /etc/nginx/streams-enabled/*.conf;
}

http {
    server {
        listen 80 default_server;
        server_name _;
        return 200;
    }
}
```

and this is the related .conf for dns:
```
upstream dns {
    server 192.162.172.120:538;
    server 192.162.172.100:538;
}

server {
    listen 53;
    include /etc/nginx/conf.d/private.conf;
    proxy_protocol on;
    proxy_pass dns;
    error_log  /var/log/nginx/dns_error.log warn;
}

server {
    listen 53 udp;
    include /etc/nginx/conf.d/private.conf;
    proxy_protocol on;
    proxy_pass dns;
    error_log  /var/log/nginx/dns_error.log warn;
}
```
The included private.conf restricts access to devices from my lan.

Oh, and I have to correct myself: DNS resolution is partly working. Some devices work without problems, others report a working internet connection but nothing works.

**Reply by bcookatpcsd** [reply]
Do you see the first post in this thread?

The stanza that begins with the word stream and all of its contents?

That is applicable to what you're using this for..

It cannot be left out..

# use the stream directive
stream {
    # for DNS-over-UDP-PROXY
    server {
        listen              53 udp;
        proxy_pass          127.0.0.1:538;
        proxy_protocol      on;
    }
    server {
    # for DNS-over-TCP-PROXY
        listen              53;
        proxy_pass          127.0.0.1:538;
        proxy_protocol      on;
    }

**Reply by kinslayer1982** [reply]
Do you see my nginx.conf? The part with the stream and the include?

**Reply by stratself** [reply]
@kinslayer1982 it's be a while since I used nginx, but I think there shouldn't be anything wrong in your conf files. The nested includes should work.

What I did notice back then was some traffic congestion, and I'm actually not sure if `worker_connections` actually helped me with it. Assuming every client machine is connected to `nginx-address:53` I'd like to ask the following:

- Are there anything peculiar in the error logs?
- Does it work better if the `upstream dns` is just one server instead of two?
- Can you resolve using localhost from inside the nginx machine (or inside the nginx container if you're using docker/podman)? Try `nslookup` in burst and see if there are any slowdowns?

Hope Shreya can also give anymore insights

**Reply by kinslayer1982** [reply]
Yes, the includes work fine, have been using them from the start without problems.

- I had it set to 1024 worker connections and there were errors about it in the logs. After increasing to 4096 there are no more errors logged.
- Using only one upstream didn't change anything.
- Running in docker, yes. I can resolve domains using nslookup. What exactly do you mean by burst? I'm just repeating the command in bash for now. It IS slow. Like 3 Seconds compared to almost instant on other machines.

Edit: Ok, fixed the slow nslookup. While testing I removed the dns servers from nginx docker compose file, so it was using dockers internal dns. It's much faster again. but there are occasional timeouts when running nslookup as a loop.

Edit 2: Just set up an nginx container on the same host as the technitium dns using just the config from the first post only adjusting the ip address. This too results in occasional timeouts. And it seems like the timeouts and the errors about the missing proxy protocol header are directly related.

**Reply by kinslayer1982** [reply]
What's interesting: The amount of timed out connections/missing proxy headers is more or less identical regardless of host machine. A raspberry pi does not produce or suffer from more connection errors than a "big" server.
Aside from that I never had problems with reverse proxying http(s) connections or proxying the dns requests without use of the proxy protocol.

**Reply by stratself** [reply]
Hi,

From your information, there seems to be a problem with nginx's handling of PROXY Protocol. I don't think turning on debug logs could reveal the root cause, but there's no harm trying. I did run nginx+tdns on a Raspberry, so it might be a platform-specific performance issue too (?). nginx/nginx#717 looks somewhat similar to your case I believe.

Since I switched to Caddy-L4, the problem seems to be gone/minimized significantly. Caddy specifies `proxy_protocol v2` directly, as v2 is the only version with proper support for UDP. If you find it suitable to switch, do let me know if it solves the timeout problem

**Reply by ShreyasZare** [reply]
@kinslayer1982 Your config looks good. I just tested same config with nginx on Ubuntu server and its working well. What platform are you running nginx on? 

I would also suggest that you run tests using DNS Client on the admin panel. This will allow you to test for both UDP and TCP transport protocols. You can also run tcpdump and check how the requests are coming from nginx server. Use wireshark to view the tcpdump pcap file to see the packet in better details.

**Comment by bcookatpcsd** [comment]
Just asking.. 

Why are you DoH proxying to Technitium.. 

<img width="936" height="559" alt="image" src="https://github.com/user-attachments/assets/89d87ba9-64e2-49c4-892e-8af4a7a37128" />

This is a VPS.. 

<img width="886" height="74" alt="image" src="https://github.com/user-attachments/assets/c3f8370b-e1a6-4a27-9c3e-0776a9d0bc9c" />

(two days b/c just did updates.. )

IMHO Technitium is quite capable all by itself.. and you build msquic.. you can get h3 and quic support.. 

https://github.com/microsoft/msquic

My 0.02

(fwiw 10/7 and beyond is less b/c I put dnsdist in front b/c I'm going to get to other Technitium back-ends.. )

**Reply by kinslayer1982** [reply]
Nobody said anything about DoH... I just mentioned that nginx never had any problems with http/https connections or dns (udp) without proxy protocol.
Or are you referring to the examples that just say that it can be done?
And there are enough reasons, load balancing multiple backends (like you mentioned yourself) or only exposing a single device/service to the internet instead of many...

Couldn't get dnsdist to connect to technitium and haproxy doesn't support udp, so I'm still trying to get nginx to work reliably.

**Reply by bcookatpcsd** [reply]
Sorry.. I'm confused 

What is nginx for?

What does Technitium and proxy protocol have to do with nginx, if not DoH..

**Reply by ShreyasZare** [reply]
@bcookatpcsd nginx is also a load balancer which can work with UDP and TCP protocols. The PROXY protocol support provides the original client IP address to the upstream.

**Comment by kinslayer1982** [comment]
> Hi,
> 
> From your information, there seems to be a problem with nginx's handling of PROXY Protocol. I don't think turning on debug logs could reveal the root cause, but there's no harm trying. I did run nginx+tdns on a Raspberry, so it might be a platform-specific performance issue too (?). [nginx/nginx#717](https://github.com/nginx/nginx/issues/717) looks somewhat similar to your case I believe.
> 
> Since I switched to Caddy-L4, the problem seems to be gone/minimized significantly. Caddy specifies `proxy_protocol v2` directly, as v2 is the only version with proper support for UDP. If you find it suitable to switch, do let me know if it solves the timeout problem

As the behavior is identical on both ARM and AMD I don't think it's platform related.
I might try Caddy, but only if I have to.

> @kinslayer1982 Your config looks good. I just tested same config with nginx on Ubuntu server and its working well. What platform are you running nginx on?
> 
> I would also suggest that you run tests using DNS Client on the admin panel. This will allow you to test for both UDP and TCP transport protocols. You can also run tcpdump and check how the requests are coming from nginx server. Use wireshark to view the tcpdump pcap file to see the packet in better details.

Almost everything here is Debian based. Proxmox, Docker VM, Nginx Docker Container, Pi OS...
I already did a tcpdump, but it doesn't tell me anything except that some requests are being sent without proxy protocol header.

Just checked again and each timeout when using a burst of nslookup requests corresponds to one missing proxy protocol header message. Which makes sense as it's only listening for proxy protocol requests on this port.
Using the client from the admin interface I do not get any errors, neither for udp nor tcp. But I am unable to send requests as fast via the gui as a shell script.

Edit: I set up a Ubuntu server VM for testing, running Nginx in Docker and directly on the VM. Same results, occasional timeouts caused by missing proxy protocol header.

**Comment by kinslayer1982** [comment]
> Managed to spin up an equivalent [Caddy-L4](https://github.com/mholt/caddy-l4/) config, for anybody interested:
> 
> ```Caddyfile
> {
>     debug
>     layer4 {
>         udp/:53 {
>             @udns dns
> 			route @udns {
> 				proxy {
>                     proxy_protocol v2
>                     upstream udp/127.0.0.1:538
>                 }
>             }
>         }
>         tcp/:53 {
> 			@tdns dns
> 			route @tdns {
> 				proxy {
>                     proxy_protocol v2
>                     upstream tcp/127.0.0.1:538
>                 }
>             }
>         }
>     }
> }
> ```
> 
> Huge thanks to people who resolved [mholt/caddy-l4#269](https://github.com/mholt/caddy-l4/issues/269)

Now it's getting interesting. I thought I'd give Caddy a try (if only to see how it behaves in comparison) and I see the exact same errors in caddy that are mentioned in this issue:
```
DBG ts=1760537914.4354799 logger=layer4 msg=matching remote=192.162.172.103:57510 error=consumed all prefetched bytes matcher=layer4.matchers.dns matched=false

DBG ts=1760537914.4355364 logger=layer4 msg=prefetched remote=192.162.172.103:57510 bytes=27
```

BUT: No more errors in Technitium DNS's logs for now.

**Comment by kinslayer1982** [comment]
@stratself
I wanted to give a little update on the Nginx problems. Did a bit more testing and opened an issue on the Nginx repo. It seems the problem is being caused by not telling Nginx to reuse the udp port 53.
Changing `listen              53 udp;` to `listen              53 udp reuseport;` I can hit it with thousands of nslookup queries and there are no timeouts and no errors about missing proxy protocol header.
Maybe you want to add that to the example config.

**Reply by stratself** [reply]
Thanks for this, I updated the initial post with your solution

**Reply by ShreyasZare** [reply]
Thanks for share it here!

**Comment by zbalkan** [comment]
Instead of dnsdist, nginx, caddy or HAproxy, can we just use another Technitium DNS Server instance, setting the resolver instances as the forwarder?

**Reply by ShreyasZare** [reply]
The PROXY protocol support is only available for inbound requests so any such setup will lose the original client IP of the request.

**Reply by zbalkan** [reply]
So, do you think there's a value in implementation of RFC 5625? In theory, it'd be a cluster of forwarders and separate proxy instances. If they all can be managed centrally using clustering, then, it might be less maintenance burden than maintaining load balancers like dnsdist, nginx, caddy, etc. "Is it worth the complexity" is the question then.

**Reply by ShreyasZare** [reply]
Its not really feasible to implement DNS proxy support since its a very specific role and not something a DNS server should implement. Using DNS server as a proxy too is not recommended since DNS server implements caching while a proxy is expected to just pass all requests to upstream.

Load balancers like nginx, etc. are recommended along with load balancer features that come with some routers/firewalls. These setup do not require frequent changes and only require change when you have to update the upstream addresses.
